library/core/src/intrinsics/simd.rs

```diff
@@ -243,12 +243,13 @@ extern "platform-intrinsic" {
     ///
     /// `T` must be a vector.
     ///
-    /// `U` must be a vector of pointers to the element type of `T`, with the same length as `T`.
+    /// `U` must be a pointer to the element type of `T`
     ///
     /// `V` must be a vector of integers with the same length as `T` (but any element size).
     ///
     /// For each element, if the corresponding value in `mask` is `!0`, read the corresponding
-    /// pointer from `ptr`.
+    /// pointer offset from `ptr`.
+    /// The first element is loaded from `ptr`, the second from `ptr.offset(1)` and so on.
```

Comment:
Is this really `offset` or `wrapping_offset`? IOW, is it UB if the `ptr` is out-of-bounds but the indices actually being access shift the pointer in-bounds?

Reply:
If I understand correctly, this is immediate UB and flagged by Miri:

```rust
let data = vec![1u32, 2, 3, 4];
let ptr = data.as_ptr().offset(-1); // UB here
```

> Both the starting and resulting pointer must be either in bounds or one byte past the end of the same [allocated object](https://doc.rust-lang.org/std/ptr/index.html#allocated-object).

But it is valid to do something like this:

```rust
let data = vec![1u32, 2, 3, 4];
let ptr = data.as_ptr().wrapping_offset(-1);
// u32x2 is a portable_simd-like #[repr(simd)] struct
let mask = u32x2::from_array([0, !0]);
let default = u32x2::from_array([111, 222]);
let v: u32x2 = simd_masked_load(mask, ptr, default);
assert_eq!(v, u32x2::from_array([111, 1]));
```

Because this does not cause an OOB read and pointer provenance is fine (?). First element would have been OOB but is not read (masked-out), second one is loaded from the first item of the Vec.

So I think you're right that it should be `wrapping_offset`

Reply:
> But it is valid to do something like this:

That is exactly the question.

If this is allowed, and LLVM also agrees it is allowed, then yes it should be `wrapping_offset`.

Reply:
@calebzulawski can you confirm what the rules should be here?
I assume there is no inbounds/nowrap requirement for the "internal" arithmetic in this operation, so it should say `wrapping_offset`.

Reply:
As far as I can tell there's no requirement, so yes, like `wrapping_offset`.

Reply:
I opened https://github.com/llvm/llvm-project/pull/77449 to clarify this on the LLVM side, but I agree with your interpretation.